File: src/main/scala/com/reddit/recommender/kafka/SimpleKafkaProducer.java
//package com.reddit.recommender.kafka;
//
//import org.apache.kafka.clients.producer.*;
//import org.json.JSONObject;
//import java.util.Properties;
//
//public class SimpleKafkaProducer implements AutoCloseable {
//    private final KafkaProducer<String, String> producer;
//
//    public SimpleKafkaProducer() {
//        Properties props = KafkaConfig.getProducerProperties();
//        this.producer = new KafkaProducer<>(props);
//    }
//
//    public void sendMessage(String topic, String key, JSONObject message) {
//        ProducerRecord<String, String> record =
//                new ProducerRecord<>(topic, key, message.toString());
//
//        producer.send(record, (metadata, exception) -> {
//            if (exception != null) {
//                System.err.println("Failed to send to " + topic + ": " + exception.getMessage());
//            }
//        });
//    }
//
//    public void sendPost(JSONObject post) {
//        String key = post.optString("id", "unknown");
//        sendMessage(KafkaConfig.getPostsTopic(), key, post);
//    }
//
//    public void sendComment(JSONObject comment) {
//        String key = comment.optString("id", "unknown");
//        sendMessage(KafkaConfig.getCommentsTopic(), key, comment);
//    }
//
//    public void flush() {
//        producer.flush();
//    }
//
//    @Override
//    public void close() {
//        producer.close();
//    }
//}

File: src/main/scala/com/reddit/recommender/kafka/SimpleKafkaProducer.scala
package com.reddit.recommender.kafka
import org.apache.kafka.clients.producer._
import org.json.JSONObject

class SimpleKafkaProducer extends AutoCloseable {
  private val producer: KafkaProducer[String, String] = new KafkaProducer[String, String](KafkaConfig.getProducerProperties())

  def sendMessage(topic: String, key: String, message: JSONObject): Unit = {
    val record = new ProducerRecord[String, String](topic, key, message.toString)

    producer.send(record, (_: RecordMetadata, e: Exception) =>
      Option(e).foreach(ex => System.err.println(s"Failed to send to $topic: ${ex.getMessage}"))
    )
  }

  def sendPost(post: JSONObject): Unit = {
    val key = Option(post.optString("id")).filter(_.nonEmpty).getOrElse("unknown")
    sendMessage(KafkaConfig.getPostsTopic(), key, post)
  }

  def sendComment(comment: JSONObject): Unit = {
    val key = Option(comment.optString("id")).filter(_.nonEmpty).getOrElse("unknown")
    sendMessage(KafkaConfig.getCommentsTopic(), key, comment)
  }

  def flush(): Unit = producer.flush()
  
  override def close(): Unit = producer.close()
}
